swift/extractor/trap/TrapLabelStore.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <cassert>
+#include <optional>
+#include <unordered_map>
+
+#include <swift/AST/ASTVisitor.h>
+
+#include "swift/extractor/trap/TrapLabel.h"
+#include "swift/extractor/trap/TrapTagTraits.h"
+// autogenerated file
+#include "swift/extractor/trap/TrapTags.h"
+
+namespace codeql {
+
+// the following is needed to avoid the problem of subclass pointers not necessarily coinciding
+// with superclass ones in case of multiple inheritance
+inline const void* getCanonicalPtr(const swift::Decl* e) {
```

Comment:
We did use `std::variant` at some point, but then changed it. @redsun82 what was the rationale?

Reply:
good point about changing the return type.

Yes, I think later we can switch to `std::variant`, or the functionally equivalent but space optimized `llvm::PointerUnion`. I'd do that once we more or less settle on all the entities we need to store (files and modules are probably the next in line)

Reply:
FWIW I also thought we needed to extract [attributes](https://github.com/github/semmle-code/pull/42508) a couple of days ago (which required another overload of `getCanonicalPtr`). Turns out this wasn't actually needed for the QL work anyway, and I closed the PR in favor of a much [simpler solution](https://github.com/github/semmle-code/pull/42530) that didn't involve another visitor. I can imagine we'll need to extract them in the future, though.

Reply:
The rationale was that it was a pain to add new types :laughing: it might probably be done so that it's less painful, but any way that's why I'm saying we should do this when the list of types we want gets stable.